lib/dynabute/dynabutable.rb

```diff
@@ -36,18 +37,73 @@ def dynabute_values
       end
 
       def dynabute_value(name: nil, field_id: nil, field: nil)
-        field = find_field(name, field_id, field)
-
-        if field.has_many
-          send(Util.value_relation_name(field.value_type)).select{|v| v.field_id == field.id }
+        field_obj = find_field(name, field_id, field)
+        field_values = send(Util.value_relation_name(field_obj.value_type))
+        field_value = if field_obj.has_many
+          field_values.select{ |v| v.field_id == field_obj.id }
         else
-          send(Util.value_relation_name(field.value_type)).detect{|v| v.field_id == field.id }
+          field_values.detect{ |v| v.field_id == field_obj.id }
         end
+        field_value
       end
 
       def build_dynabute_value(name: nil, field_id: nil, field: nil, **rest)
-        field = find_field(name, field_id, field)
-        send(Util.value_relation_name(field.value_type)).build(field_id: field.id, **rest)
+        field_obj = find_field(name, field_id, field)
+        send(Util.value_relation_name(field_obj.value_type)).build(field_id: field_obj.id, **rest)
+      end
+
+      # Returns value attribute for specified field.
+      # If field can have multiple values for single target model object,
+      # an array of values is returned, unless specific value_id is provided.
+      def get_dynabute_value(name: nil, field_id: nil, field: nil, value_id: nil)
+        field_obj = find_field(name, field_id, field)
+        value_obj = dynabute_value(field: field_obj)
+        return unless value_obj
+        if field_obj.has_many && value_id
+          value_obj = value_obj.detect{|v| v.id == value_id}
+        end
+        value_obj.is_a?(Array) ? value_obj.map(&:value) : value_obj&.value
+      end
+
+      # Sets the value in the target model object nested attribute structure.
+      # If field can have multiple values for single target model object,
+      # a new value will be added, unless specific value_id os provided.
+      #
+      # This method does not store changes in the database. "save" method should
+      # be called on target model to store all changes, or individually on every
+      # value record returned by this method.
+      def set_dynabute_value(name: nil, field_id: nil, field: nil, value: nil, value_id: nil)
+        field_obj = find_field(name, field_id, field)
+        value_obj = dynabute_value(field: field_obj)
+        if field_obj.has_many
+          if value_id
+            value_obj = value_obj.detect{|v| v.id == value_id} if value_obj
+            fail ValueNotFound unless value_obj
+          else
+            value_obj = build_dynabute_value(field: field_obj)
+          end
```

Comment:
set_dynabute_attributes() for has_many field sounds more like it's replacing the whole values with the given  array. having `append_dynabute_value` (which  should throw error when called on non has_many field) seems more natural to me, what do you think?

Reply:
My idea was that you have single method for setting the values. It was intended to be like `create_or_update` concept.
Just like rails, when submitting a form for persisted object (`id` exists), it will trigger update action. In case the object is not persisted and there is no `id` value, form will submit create action and the new record will be added to the database.

So first time you call method for `has_many` field, it will add specified value, same as with strings and integers, but it will return an Array. Next set will just append to that Array as something already exists.

Let me know if you have something specific on your mind 😉